Support For Energy in pyatmo

tmenguy opened the pull request:
Hi,

This one is the companion of https://github.com/tmenguy/netatmo_custom. 
All of this is to add energy measure support in homeassitant, so for pyatmo, extended handling of the /getmeasure endpoint. This one is not really up to date by the hour, from what I've seen it is updated very 3 to 4 hours, so I introduced an estimation using power and the measure to have a proper following of energy in homeassitant that is made to be "close to real time" .

Of course feel free to ask questions and so on. What I don't know is : is this one is working with the existing home assistant integration, it should, it is fixing some issues, but well, we never know, and I don't have a full HA setup for dev, more one for HACS.

BR
Thomas

cgtobi commented on src/pyatmo/account.py:
```diff
@@ -31,11 +33,17 @@
 
         self.auth: AbstractAsyncAuth = auth
         self.user: str | None = None
+        self.support_only_homes = support_only_homes
+        self.all_account_homes: dict[str, Home] = {}
+        self.additional_public_homes: dict[str, Home] = {}
```
What do these mean, can you please elaborate?

tmenguy replied:
hi, first thx for review: I have added the notion of "selected homes" but wanted to keep everything the same vs the previous implementation, hence keeping in separate bucket the different types of homes we now have.

cgtobi commented on src/pyatmo/account.py:
```diff
@@ -49,17 +57,43 @@ def __repr__(self) -> str:
             f"{self.__class__.__name__}(user={self.user}, home_ids={self.homes.keys()}"
         )
 
+    def update_supported_homes(self, support_only_homes: list | None = None):
```
What is `support_only_homes`?

tmenguy replied:
see above : as set of homes that you define in your option config : see the readme I tried to explain :) this variable is for th list of homes the user decided to "import" or support in his HA instance. Ex : if you have a main home and a secondary vacation house, you probably don't want to have your vacation house in your main house HA instance

cgtobi commented on src/pyatmo/modules/module.py:
```diff
@@ -621,55 +600,333 @@ def __init__(self, home: Home, module: ModuleT):
+        else:
+            return len(self.home.energy_endpoints)
 
     async def async_update_measures(
```
This function is way to big and deeply nested. If it actually needs to do all that this needs to be cut into manageable and testable functions.

tmenguy replied:
are you talking about async_update_measures?

tmenguy replied:
ok it is big, will refactor a bit, but it will be artificial to break in small pieces ... I personally prefer reading everything at once :)

cgtobi commented on src/pyatmo/modules/module.py:
```diff
@@ -621,55 +600,333 @@ def __init__(self, home: Home, module: ModuleT):
+        self.sum_energy_elec_peak = 0
+        self.sum_energy_elec_off_peak = 0
+
+    def get_sum_energy_elec_power_adapted(self, to_ts: int | float | None = None, conservative: bool = False):
```
This code is too nested, which makes it hard to read, hard to test and error prone.

tmenguy replied:
are you talking about get_sum_energy_elec_power_adapted ?

tmenguy replied:
I'll refactor a bit (even if this one is honestly not so big)

cgtobi reviewed (commented):
Thank you for your effort. I feel that this makes the code quite hard to follow and I am not sure if this solves the energy tracking properly. Netatmo has been quite sensitive to aggressive polling and I don't think we actually need this. No matter if you fetch data ever minute or every 10 seconds, it will be far off from realtime. Please correct me if I misinterpret your intention. Also I think this PR tries to solve too many things at once, which makes reviewing quite difficult.

tmenguy commented:
> Thank you for your effort. I feel that this makes the code quite hard to follow and I am not sure if this solves the energy tracking properly. Netatmo has been quite sensitive to aggressive polling and I don't think we actually need this. No matter if you fetch data ever minute or every 10 seconds, it will be far off from realtime. Please correct me if I misinterpret your intention. Also I think this PR tries to solve too many things at once, which makes reviewing quite difficult.

Hi ... yes I do think you missed some stuff here :)
-there is the addition of the support of multi home support ... I'm not the only one with the need I think
-previously the historical data and everything related to energy was really not working : as it was not supporting off peak / peak mode (very popular in France)
-as for the polling : I actually made a big rework here has now the throttling is mathematically correct, and adapt gracefully with the number of publisher to be probed, it was mandatory as for energy you have to query each device individually : I probe then every 30mn, but depending of the number of devices, I adapt this throittling to always comply with Netatmo given number. I also very finely manage the exact number of calls per hour in a floating 1h window to be sure I stay in (and actually below Netatmo recommendation, I've introduced also a proper handling of their throttling error if even fired, and a proportional/Derivative adaptation of the throttling.
Also the energy APIs from Netatmo seems rather randomly being updated sometimes every 30mn, sometimes every 4hours, I also introduced and adaptation using the power data obtained to approximate the energy value during those non update period of time

All in all yes the code is more complexe but:
-the throttling is a lot more robust and predictable than before
-the peak off peak model of Netatmo is a nightmare to handle, but we have it now
-the energy measure are the best we can get.

I have a massive installation of Legrand devices (90 measure points) and it is working quite well, no throttling from the API (but well yes some timeout but it is different)

Hope it helps to understand the work here :)

BR
Thomas

tmenguy commented:
Small nudge :)

cgtobi commented on src/pyatmo/account.py:
```diff
@@ -49,17 +60,38 @@ def __repr__(self) -> str:
+    def update_supported_homes(self, support_only_homes: list | None = None):
+        """Update the exposed/supported homes."""
+
+        if support_only_homes is None or len(support_only_homes) == 0:
```
`if not support_only_homes` should be sufficient, no?

cgtobi commented on src/pyatmo/account.py:
```diff
@@ -70,51 +102,82 @@ async def async_update_topology(self) -> None:
 
-    async def async_update_events(self, home_id: str) -> None:
-        """Retrieve events from /getevents."""
+    async def async_update_status(self, home_id: str | None = None) -> int:
+        """Retrieve status data from /homestatus. Returns the number of performed API calls."""
+
+        if home_id is None:
+            self.update_supported_homes(self.support_only_homes)
+            homes = self.homes
+        else:
+            homes = [home_id]
+        num_calls = 0
+        all_homes_ok = True
+        for h_id in homes:
+            resp = await self.auth.async_post_api_request(
+                endpoint=GETHOMESTATUS_ENDPOINT,
+                params={"home_id": h_id},
+            )
+            raw_data = extract_raw_data(await resp.json(), HOME)
+            is_correct_update = await self.homes[h_id].update(raw_data)
+            if not is_correct_update:
+                all_homes_ok = False
+            num_calls += 1
+
+        if all_homes_ok is False:
+            raise ApiHomeReachabilityError(
+                "No Home update could be performed, all modules unreachable and not updated",
+            )
+
+        return num_calls
```
This looks unnecessarily complex compared to the original method. I also don't like the passing and counting of the calls. I'm sure there is a more elegant way of doing this.

tmenguy replied:
well it was to add support for an "all homes" status update. The num call counts is something I have added everywhere for consistency to allow code using this lib to do a correct job at assessing the number of calls made to nettle/legrand (as we, rightfully, don't handle throttling at the pyatmo lib level

tmenguy replied:
it introduces also reachability error handling

cgtobi commented on src/pyatmo/account.py:
```diff
@@ -70,51 +102,82 @@ async def async_update_topology(self) -> None:
         days: int = 7,
-    ) -> None:
-        """Retrieve measures data from /getmeasure."""
+        end_time: int | None = None,
```
Let's move this right after `start_time`.

tmenguy replied:
it was to not break compatibilities here ... but if ok for you yep sure :)

cgtobi commented on src/pyatmo/account.py:
```diff
@@ -49,17 +60,38 @@ def __repr__(self) -> str:
             f"{self.__class__.__name__}(user={self.user}, home_ids={self.homes.keys()}"
         )
 
+    def update_supported_homes(self, support_only_homes: list | None = None):
```
Why is support_only_homes needed as a parameter when it gets set at init time?
